File: 0160 - Intersection of Two Linked Lists/Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        first, second = headA, headB
        len_first, len_second = 0, 0
        while first is not None:
            first = first.next;
            len_first = len_first + 1
            
        while second is not None:
            second = second.next;
            len_second = len_second + 1
            
        first, second = headA, headB
        
        if len_second > len_first:
            dif = len_second - len_first
            for i in range(dif):
                second = second.next
        else:
            dif = len_first - len_second
            for i in range(dif):
                first = first.next
            
        while first != second:
            first = first.next
            second = second.next
        
        return first

File: 0160 - Intersection of Two Linked Lists/Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* first = headA;
        ListNode* second = headB;
        int len_first = 0, len_second = 0;
        while (first != NULL) {
            ++len_first;
            first = first->next;
        }
        while (second != NULL) {
            ++len_second;
            second = second->next;
        }
        first = headA;
        second = headB;
        if (len_first > len_second) {
            int dif = len_first - len_second;
            for (int i = 0; i < dif; ++i) {
                first = first->next;
            }
        } else {
            int dif = len_second - len_first;
            for (int i = 0; i < dif; ++i) {
                second = second->next;
            }
        }
        while (first != second) {
            first = first->next;
            second = second->next;
        }
        return first;
    }
};
